### PR TITLE
Use a resource-based installation method for Nexus3

### PR DESCRIPTION
**Pull request by olive42**
Make it easy to install several instances of Nexus on the same server. Usage:
nexus3 'instance_name' do
  action :install # default
end

Add unit and integration tests that work with a Vagrant and Docker provider.
Documentation update.

### CODE REVIEW

File: resources/service_systemd.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    provider Chef::Provider::Service::Systemd
+    supports status: true
+    action :enable
+    only_if { ::File.exist?("/etc/systemd/system/nexus3_#{new_resource.instance_name}.service") }
+  end
+end
+
+action_class do
+  def create_init
+    template "/etc/systemd/system/nexus3_#{new_resource.instance_name}.service" do
```
Comment (jmauro) [inline resources/service_systemd.rb]: You have to create instance systemd service, if I am not clear, I would be glad to show an example

File: resources/service_sysvinit.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    action :start
+  end
+end
+
+action :stop do
+  service "nexus3_#{new_resource.instance_name}" do
+    provider platform_sysv_init_class
+    supports status: true
+    action :stop
+    only_if { ::File.exist?("/etc/init.d/nexus3_#{new_resource.instance_name}") }
```
Comment (jmauro) [inline resources/service_sysvinit.rb]: Since you want to `stop` the service is not declared?

File: resources/service_sysvinit.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    supports status: true
+    action :restart
+  end
+end
+
+action :enable do
+  create_init
+
+  service "nexus3_#{new_resource.instance_name}" do
+    provider platform_sysv_init_class
```
Comment (jmauro) [inline resources/service_sysvinit.rb]: why do you need this?

File: resources/service_sysvinit.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    supports status: true
+    action :enable
+    only_if { ::File.exist?("/etc/init.d/nexus3_#{new_resource.instance_name}") }
```
Comment (jmauro) [inline resources/service_sysvinit.rb]: same as the action `stop`

File: resources/service_sysvinit.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+action_class do
+  def create_init
+    link ::File.join('/etc/init.d', "nexus3_#{new_resource.instance_name}") do
+      to ::File.join(new_resource.install_dir.to_s, 'bin', 'nexus')
+      owner new_resource.nexus3_user
+      group new_resource.nexus3_group
+    end
+
+    if platform_family?('rhel', 'fedora')
```
Comment (jmauro) [inline resources/service_sysvinit.rb]: enable should do that

File: resources/service_sysvinit.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        command "chkconfig --levels 345 nexus3_#{new_resource.instance_name} on"
+      end
+    else
+      execute 'update-rc.d' do
+        command "update-rc.d nexus3_#{new_resource.instance_name} defaults"
+      end
+    end
+  end
+
+  def platform_sysv_init_class
```
Comment (jmauro) [inline resources/service_sysvinit.rb]: why is this needed?

File: templates/default/init_systemd.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+[Unit]
```
Comment (jmauro) [inline templates/default/init_systemd.erb]: Like I said above maybe a more generic file could be used (to be confirmed)

File: test/fixtures/cookbooks/nexus3_test/recipes/default.rb [hunk 8]
```diff
@@ -1,2 +1,37 @@
 include_recipe 'java_se'
-include_recipe 'nexus3'
+
+# The default Kitchen image may not be up-to-date.
+apt_update 'update' do
```
Comment (jmauro) [inline test/fixtures/cookbooks/nexus3_test/recipes/default.rb]: What!!! Debian specific...

File: test/fixtures/cookbooks/nexus3_test/recipes/default.rb [hunk 9]
```diff
@@ -1,2 +1,37 @@
...
+  action :update
+end
+
+# On Debian-style systems, curl is not installed on the kitchen
+# images. Install it here since we only need it for testing.
+package 'curl' do
+  only_if { platform_family?('debian') }
```
Comment (jmauro) [inline test/fixtures/cookbooks/nexus3_test/recipes/default.rb]: We don't care about this constraint just install it

File: attributes/default.rb [hunk 10]
```diff
@@ -1,25 +1,28 @@
...
-  default['nexus3']['path'] = ENV['SYSTEMDRIVE']
+  default['nexus3']['path'] = 'C:\Nexus3'
+  default['nexus3']['group'] = 'Administrators'
 else
-  default['nexus3']['url'] = 'http://download.sonatype.com/nexus/3/latest-unix.tar.gz'
   default['nexus3']['path'] = '/opt'
+  default['nexus3']['group'] = 'nexus'
 end
+# Download URL is defined in the resource but you can override it with the default['nexus3']['url'] attribute
+default['nexus3']['url'] = nil
```
Comment (olive42) [inline attributes/default.rb]: Fixed. Do you want to also include the matching sha256sum or should I leave nil as optional?

File: resources/default.rb [hunk 11]
```diff
@@ -1,18 +1,170 @@
...
+    shell '/bin/bash'
+  end
+
+  group new_resource.nexus3_group do
+    members nexus3_user
+    append true
+  end
+
+  # Install Nexus3 software
+  include_recipe 'ark'
```
Comment (olive42) [inline resources/default.rb]: Moved include_recipe to the default recipe. I even had read this bug in a different context not long ago :)

File: resources/default.rb [hunk 12]
```diff
@@ -1,18 +1,170 @@
...
+
+  # Install configuration from templates
+  template ::File.join(install_dir, 'bin', 'nexus.rc') do
+    source 'nexus.rc.erb'
+    variables(user: nexus3_user)
+    mode '0644'
+    owner nexus3_user
+    group nexus3_group
+    cookbook 'nexus3'
+    notifies(:write, "log[#{instance_name} is restarting]", :delayed)
```
Comment (olive42) [inline resources/default.rb]: Sounds good, although I don't want to restart the service until all files are in place, so :delayed seems the better way.

File: resources/default.rb [hunk 13]
```diff
@@ -1,18 +1,170 @@
...
+  end
+
+  vmoptions = {}
+  vmoptions.merge!(vmoptions_variables)
```
Comment (olive42) [inline resources/default.rb]: This is to merge in 'data' if not provided in the vmoptions. Alternatively, we can force providing it in the vmoptions Hash and remove the :data property.

File: resources/default.rb [hunk 14]
```diff
@@ -1,18 +1,170 @@
...
+  end
+
+  nexus3_service instance_name.to_s do
+    install_dir install_dir
+    nexus3_user new_resource.nexus3_user
+    nexus3_group new_resource.nexus3_group
+    action :enable
+  end
+
+  log "#{new_resource.instance_name} is restarting" do
```
Comment (olive42) [inline resources/default.rb]: Removed.

File: resources/default.rb [hunk 15]
```diff
@@ -1,18 +1,170 @@
...
+    url << if platform?('windows')
+             '-win64.zip'
+           else
+             '-unix.tar.gz'
+           end
+    url
+  end
+
+  # Raise an error in case Nexus takes a really long time to start.
+  class Nexus3NotReady < StandardError
```
Comment (olive42) [inline resources/default.rb]: It used to be in a library (in the recipe mode). I can put it back there.

File: resources/service_systemd.rb [hunk 16]
```diff
@@ -0,0 +1,82 @@
+provides :nexus3_service_systemd
+
+provides :nexus3_service, os: 'linux' do |_node|
+  Chef::Platform::ServiceHelpers.service_resource_providers.include?(:systemd)
```
Comment (olive42) [inline resources/service_systemd.rb]: Done

File: resources/service_systemd.rb [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+property :instance_name, String, name_property: true
+property :install_dir, String
+property :nexus3_user, String
+property :nexus3_group, String
+
+action :start do
+  create_init
+
+  service "nexus3_#{new_resource.instance_name}" do
+    provider Chef::Provider::Service::Systemd
```
Comment (olive42) [inline resources/service_systemd.rb]: I don't know. This is hitting the limit of my understanding.

File: resources/service_systemd.rb (hunk 1 above)
Comment (olive42) [inline resources/service_systemd.rb]: The action_class doc at https://docs.chef.io/custom_resources.html didn't make it look like it was necessary, so I probab;y misunderstand something.

File: resources/service_sysvinit.rb (hunk 3 above)
Comment (olive42) [inline resources/service_sysvinit.rb]: Mostly I got "copy-n-paste inspiration" from the tomcat cookbook. Maybe they're wrong.

File: test/fixtures/cookbooks/nexus3_test/recipes/default.rb (hunk 8 above)
Comment (olive42) [inline test/fixtures/cookbooks/nexus3_test/recipes/default.rb]: The beauty of it is that it doesn't care or run on non-Debian platforms.

File: test/fixtures/cookbooks/nexus3_test/recipes/default.rb (hunk 9 above)
Comment (olive42) [inline test/fixtures/cookbooks/nexus3_test/recipes/default.rb]: Ok.

File: attributes/default.rb (hunk 10 above)
Comment (jmauro) [inline attributes/default.rb]: We should just provide a default setting for people to use easily

File: resources/default.rb (hunk 11 above)
Comment (jmauro) [inline resources/default.rb]: And have to make advice people to include the recipe default when using the resource (I think the only problem is on Windows, Linux just install a package so we can just include the package but let's talk about it if you want)

File: resources/default.rb (hunk 12 above)
Comment (jmauro) [inline resources/default.rb]: Why not

File: resources/default.rb (hunk 13 above)
Comment (jmauro) [inline resources/default.rb]: I am sorry I don't understand the explanation

File: resources/service_systemd.rb (hunk 1 above)
Comment (jmauro) [inline resources/service_systemd.rb]: I was talking about systemd instance not chef instance

File: test/fixtures/cookbooks/nexus3_test/recipes/default.rb (hunk 8 above)
Comment (jmauro) [inline test/fixtures/cookbooks/nexus3_test/recipes/default.rb]: why need an update system?

File: resources/default.rb (hunk 13 above)
Comment (olive42) [inline resources/default.rb]: Ok I found a much better way of using vmoptions_variable and merging in the data resource:
template blah do
  variables(
    vmoptions_variables.merge(data: new_resource.data)
  )
end

File: resources/service_systemd.rb (hunk 1 above)
Comment (olive42) [inline resources/service_systemd.rb]: This will be in another commit in this branch. Soon!

It is a bit convoluted to use version as the %i in systemd in the general case, but why not. I think we should probably use another identifier that people can set to their own liking.

As to using "service nexus3@3.2.1-01", this is messy on non-systemd systems.

Comment (jmauro) [inline resources/service_systemd.rb]: I am only talking about systemd services, like I said the version was just an idea.

Comment (olive42) [inline resources/service_systemd.rb]: Ok Systemd is starting to get really annoying. It will not accept a variable as its command for ExecStart / ExecStop. Using ExecStart=/bin/sh -c $START_COMMAND instead + a User=Nexus fails to start the service. I have not found a way that starts it properly apart from having ExecStart=/opt/nexus-3.3.1-01/bin/nexus start in the service file, which renders this whole thing useless.

I suggest to make it simple and create a unit file for every instance of Nexus we want to bring up. I suspect it will not really be a big issue (if we find someone willing to bring up more than 3 different instances of Nexus on the same machine, we can offer them all our sympathies and suggest changing careers).

**Comment by olive42** [comment]
Ping on this?

File: resources/default.rb [hunk 18]
```diff
@@ -1,18 +1,125 @@
...
+
+  user new_resource.nexus3_user do
+    comment 'Nexus 3 user'
+    home nexus3_home
+    manage_home false # is linked to install_dir below
+    shell '/bin/bash'
+  end
+
+  group new_resource.nexus3_group do
+    members nexus3_user
```
Comment (jmauro) [inline resources/default.rb]: Same a previous comment

File: resources/default.rb [hunk 19]
```diff
@@ -1,18 +1,125 @@
...
+    path new_resource.path
+    owner new_resource.nexus3_user
+    group new_resource.nexus3_group
+  end
+
+  # Setup directories
+  [install_dir, data, ::File.join(data, 'etc')].each do |dir|
+    directory dir do
+      recursive true
+      owner nexus3_user
```
Comment (jmauro) [inline resources/default.rb]: Use `new_resource`

File: resources/default.rb [hunk 20]
```diff
@@ -1,18 +1,125 @@
...
+      group nexus3_group
```
Comment (jmauro) [inline resources/default.rb]: Plz use `new_resource`

File: resources/default.rb [hunk 21]
```diff
@@ -1,18 +1,125 @@
...
+      mode '0755'
+    end
+  end
+
+  # Install configuration from templates
+  template ::File.join(install_dir, 'bin', 'nexus.rc') do
+    source 'nexus.rc.erb'
+    variables(user: nexus3_user)
```
Comment (jmauro) [inline resources/default.rb]: same as previous comment

File: resources/default.rb [hunk 22]
```diff
@@ -1,18 +1,125 @@
...
+    mode '0644'
+    owner nexus3_user
```
Comment (jmauro) [inline resources/default.rb]: Same as previous comment

File: resources/default.rb [hunk 23]
```diff
@@ -1,18 +1,125 @@
...
+    group nexus3_group
```
Comment (jmauro) [inline resources/default.rb]: Same as previous comment

File: resources/default.rb [hunk 24]
```diff
@@ -1,18 +1,125 @@
...
+    cookbook 'nexus3'
+    notifies :restart, "nexus3_service[#{new_resource.service_name}]", :delayed
+    notifies :create, 'ruby_block[block until operational]', :delayed
+  end
+
+  template ::File.join(install_dir, 'bin', 'nexus.vmoptions') do
```
Comment (jmauro) [inline resources/default.rb]: shouldn't this be define by instance? Or is this common to all nexus instance?

File: resources/service_systemd.rb (hunk 1 above)
Comment (jmauro) [inline resources/service_systemd.rb]: If so then use the systemd cookbook to create unit

File: attributes/default.rb [hunk 25]
```diff
@@ -1,25 +1,26 @@
...
+  default['nexus3']['path'] = 'C:\Nexus3'
+  default['nexus3']['group'] = 'Administrators'
 else
-  default['nexus3']['url'] = 'http://download.sonatype.com/nexus/3/latest-unix.tar.gz'
   default['nexus3']['path'] = '/opt'
+  default['nexus3']['group'] = 'nexus'
 end
+# Download URL is defined in the resource but you can override it with the default['nexus3']['url'] attribute
+default['nexus3']['version'] = '3.2.1-01'
+default['nexus3']['url'] = 'https://download.sonatype.com/nexus/3/nexus-3.2.1-01-unix.tar.gz'
```
Comment (olive42) [inline attributes/default.rb]: Done

File: libraries/nexus_helper.rb [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+        super <<-EOH
+The Nexus server at #{endpoint} did not become ready within #{timeout}
+seconds. On large Nexus instances (or on smaller hardware), you may need
+to increase the timeout to #{timeout * 4} seconds. Alternatively, Nexus
+may have failed to start.
+EOH
+      end
+    end
+
+    def wait_until_ready!(endpoint)
```
Comment (olive42) [inline libraries/nexus_helper.rb]: Done, with a default and called with node['nexus3']['api']['wait'] (the 'wait' attribute name comes from the original cookbook and I think we should not necessarily rename it right away.)

File: resources/default.rb [hunk 27]
```diff
@@ -1,18 +1,125 @@
...
+# property :nexus3_password, kind_of: [String, NilClass], sensitive: true # Admin password
+property :version, kind_of: String, default: lazy { node['nexus3']['version'] }
+property :url, kind_of: String, default: lazy { node['nexus3']['url'] }
+property :checksum, kind_of: String, default: lazy { node['nexus3']['checksum'] }
+property :nexus3_home, kind_of: String, default: lazy { node['nexus3']['home'] }
+property :path, kind_of: String, default: lazy { node['nexus3']['path'] }
+property :data, kind_of: String, default: lazy { node['nexus3']['data'] }
+# Name of service used as property for nexus3_service; can be instance_name or something different
+# like version number.
+property :service_name, kind_of: String, default: lazy { instance_name }
```
Comment (olive42) [inline resources/default.rb]: Done

File: resources/default.rb (hunk 18 above)
Comment (olive42) [inline resources/default.rb]: done

File: resources/default.rb [hunk 28]
```diff
@@ -1,18 +1,125 @@
...
+property :properties_variables, kind_of: Hash, default: lazy { node['nexus3']['properties_variables'] }
+property :vmoptions_variables, kind_of: Hash, default: lazy { node['nexus3']['vmoptions_variables'] }
+
+action :install do
+  install_dir = ::File.join(path, "nexus-#{version}")
+
+  user new_resource.nexus3_user do
+    comment 'Nexus 3 user'
+    home nexus3_home
```
Comment (olive42) [inline resources/default.rb]: done

File: resources/default.rb [hunk 29]
```diff
@@ -1,18 +1,125 @@
...
+    shell '/bin/bash'
+  end
+
+  group new_resource.nexus3_group do
+    members nexus3_user
+    append true
+  end
+
+  # Install Nexus3 software
+  ark "nexus-#{new_resource.version}" do
```
Comment (olive42) [inline resources/default.rb]: Documentation updated.

File: resources/default.rb [hunk 30]
```diff
@@ -1,18 +1,125 @@
...
+      Chef::Log.info "Waiting until Nexus is listening on port #{node['nexus3']['properties_variables']['port']}"
+      wait_until_ready!(node['nexus3']['api']['endpoint'])
+    end
+    action :nothing
+  end
+end
+
+action_class do
+  def download_url
+    return node['nexus3']['url'] unless node['nexus3']['url'].nil? || node['nexus3']['url'].empty?
```
Comment (olive42) [inline resources/default.rb]: No good idea why, changed.

File: resources/default.rb (hunk 24 above)
Comment (olive42) [inline resources/default.rb]: Ok, used new_resource there as well :) This should probably not be shared among Nexus instances.

File: resources/default.rb (hunk 27 above)
Comment (olive42) [inline resources/default.rb]: I am not sure I understand.

If you suggest writing `property :service_name, kind_of: String, default: lazy { new_resource.instance_name }`, this is not working: ``nexus3[foo] (nexus3_test::default line 6) had an error: NoMethodError: undefined method `new_resource' for Custom resource nexus3 from cookbook nexus3``

**Comment by charlesjohnson** [comment]
I'm excited about this branch! Thanks for your hard work.

File: metadata.rb [hunk 31]
```diff
@@ -11,6 +11,7 @@
 chef_version '>= 12.14.34'
 
 depends 'ark'
+depends 'build-essential'
```
Comment (Annih) [inline metadata.rb]: is that required? I think you had a berkshelf bug. If you don't rely on build-essential directly, don't add it.

Comment (olive42) [reply metadata.rb]: Berkshelf bug was fixed, now removed this dependency.
